File: src/leetcode/python/easy/BalancedBinaryTree_110.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class BalancedBinaryTree_110(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def dfs(node):
            if not node:
                return 0
            left, right = dfs(node.left), dfs(node.right)
            if left == -1 or right == -1 or abs(left - right) > 1:
                return -1
            return max(left, right) + 1

        return dfs(root) != -1

File: src/leetcode/java/BalancedBinaryTree_110.java
package leetcode.java;

public class BalancedBinaryTree_110 {
  public boolean isBalanced(TreeNode root) {
    return dfs(root) != -1;
  }

  private static int dfs(TreeNode node) {
    if (node == null) return 0;
    int left = dfs(node.left);
    int right = dfs(node.right);
    if (Math.abs(left - right) > 1 || left == -1 || right == -1) return -1;
    return Math.max(left, right) + 1;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
